service/s3/api_client.go

```diff
@@ -452,3 +452,125 @@ func addRequestResponseLogging(stack *middleware.Stack, o Options) error {
+func WithPresignExpires(dur time.Duration) func(*PresignOptions) {
+	return withPresignExpires(dur).options
+}
+
+type withPresignExpires time.Duration
+
+func (w withPresignExpires) options(o *PresignOptions) {
+	o.Expires = time.Duration(w)
+}
+
+// PresignClient represents the presign url client
+type PresignClient struct {
+	client    *Client
+	presigner v4.HTTPPresigner
+	expires   time.Duration
+}
+
+// NewPresignClient generates a presign client using provided Client options and
+// presign options
+func NewPresignClient(options Options, optFns ...func(*PresignOptions)) *PresignClient {
+	var presignOptions PresignOptions
+	for _, fn := range optFns {
+		fn(&presignOptions)
+	}
+	client := New(options, presignOptions.ClientOptions...)
+	var presigner v4.HTTPPresigner
+	if presignOptions.Presigner != nil {
+		presigner = presignOptions.Presigner
+	} else {
+		presigner = v4.NewSigner()
+	}
+	return &PresignClient{
+		client: client, presigner: presigner,
+		expires: presignOptions.Expires,
+	}
+}
+
+// NewPresignClientWrapper generates a presign client using provided API Client and
+// presign options
+func NewPresignClientWrapper(c *Client, optFns ...func(*PresignOptions)) *PresignClient {
+	var presignOptions PresignOptions
+	for _, fn := range optFns {
+		fn(&presignOptions)
+	}
+	client := copyAPIClient(c, presignOptions.ClientOptions...)
+	var presigner v4.HTTPPresigner
+	if presignOptions.Presigner != nil {
+		presigner = presignOptions.Presigner
+	} else {
+		presigner = v4.NewSigner()
+	}
+	return &PresignClient{
+		client: client, presigner: presigner,
+		expires: presignOptions.Expires,
+	}
+}
+
+// NewPresignClientFromConfig generates a presign client using provided AWS config
+// and presign options
+func NewPresignClientFromConfig(cfg aws.Config, optFns ...func(*PresignOptions)) *PresignClient {
+	var presignOptions PresignOptions
+	for _, fn := range optFns {
+		fn(&presignOptions)
+	}
+	client := NewFromConfig(cfg, presignOptions.ClientOptions...)
+	var presigner v4.HTTPPresigner
+	if presignOptions.Presigner != nil {
+		presigner = presignOptions.Presigner
+	} else {
+		presigner = v4.NewSigner()
+	}
+	return &PresignClient{
+		client: client, presigner: presigner,
+		expires: presignOptions.Expires,
+	}
+}
+func copyAPIClient(c *Client, optFns ...func(*Options)) *Client {
+	return New(c.options, optFns...)
+}
+func (c *PresignClient) convertToPresignMiddleware(stack *middleware.Stack, options Options) (err error) {
+	stack.Finalize.Clear()
+	stack.Deserialize.Clear()
+	stack.Build.Remove((*awsmiddleware.ClientRequestID)(nil).ID())
+	err = stack.Finalize.Add(v4.NewPresignHTTPRequestMiddleware(options.Credentials, c.presigner), middleware.After)
+	if err != nil {
+		return err
+	}
+	if c.expires != 0 {
+		// add middleware to set expiration for s3 presigned url
+		err = stack.Build.Add(&s3cust.AddExpiresOnPresignedURL{Expires: c.expires}, middleware.After)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
```

Comment:
I'm not 100% positive about this, but the `X-Amz-Expires` header may be required for S3 presigned operations. We should validate this. From the Amazon S3 docs on presigned URLs it not really clear, other than saying the minimum time is 1 second. 

https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html